core/src/main/java/org/elasticsearch/search/profile/Timer.java

```diff
@@ -33,38 +33,59 @@
  *  }
  *  </pre>
  */
-// TODO: do not time every single call as discussed in https://github.com/elastic/elasticsearch/issues/24799
-public final class Timer {
+public class Timer {
 
-    private long timing, count, start;
+    private boolean doTiming;
+    private long timing, count, lastCount, start;
+
+    /** pkg-private for testing */
+    long nanoTime() {
+        return System.nanoTime();
+    }
 
     /** Start the timer. */
-    public void start() {
+    public final void start() {
         assert start == 0 : "#start call misses a matching #stop call";
+        // We measure the timing of each method call for the first 256
+        // calls, then 1/2 call up to 512 then 1/3 up to 768, etc. with
+        // a maximum interval of 1024, which is reached for 1024*2^8 ~= 262000
+        // This allows to not slow down things too much because of calls
+        // to System.nanoTime() when methods are called millions of time
+        // in tight loops, while still providing useful timings for methods
```

Comment:
I was not aware of the relative times. What I meant was that calling `System.nanoTime()` takes on typical systems 30ns (see [nanotrusting nanotime](https://shipilev.net/blog/2014/nanotrusting-nanotime/#_one_off_measurements)). If the code you want to measure takes a similarly long time then you will have a significant measurement error. However, if you measure code that takes e.g. 500µs to execute, the overhead is much lower.